src/core/ext/filters/client_channel/lb_policy/rls/rls.cc

```diff
@@ -1419,35 +1425,43 @@ void RlsLb::Cache::ResetAllBackoff() {
 void RlsLb::Cache::Shutdown() {
   map_.clear();
   lru_list_.clear();
-  grpc_timer_cancel(&cleanup_timer_);
+  if (cleanup_timer_handle_.has_value() &&
+      lb_policy_->channel_control_helper()->GetEventEngine()->Cancel(
+          *cleanup_timer_handle_)) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_lb_rls_trace)) {
+      gpr_log(GPR_INFO, "[rlslb %p] cache cleanup timer canceled", lb_policy_);
+    }
+    cleanup_timer_handle_.reset();
+    lb_policy_->Unref(DEBUG_LOCATION, "Shutdown");
+  }
 }
 
-void RlsLb::Cache::OnCleanupTimer(void* arg, grpc_error_handle error) {
-  Cache* cache = static_cast<Cache*>(arg);
-  cache->lb_policy_->work_serializer()->Run(
-      [cache, error]() {
-        RefCountedPtr<RlsLb> lb_policy(cache->lb_policy_);
+void RlsLb::Cache::OnCleanupTimer() {
+  {
+    MutexLock lock(&lb_policy_->mu_);
+    cleanup_timer_handle_.reset();
+  }
+  lb_policy_->work_serializer()->Run(
+      [this]() {
+        RefCountedPtr<RlsLb> lb_policy(lb_policy_);
         if (GRPC_TRACE_FLAG_ENABLED(grpc_lb_rls_trace)) {
-          gpr_log(GPR_INFO, "[rlslb %p] cache cleanup timer fired (%s)",
-                  cache->lb_policy_, StatusToString(error).c_str());
+          gpr_log(GPR_INFO, "[rlslb %p] cache cleanup timer fired", lb_policy_);
         }
-        if (error == absl::CancelledError()) return;
```

Comment:
Is there a similar issue here, i.e. the logic of the timer being cancelled is conflated with shutdown?

Reply:
Indeed, fixed. Great catch!